File: 0491-non-decreasing-subsequences/0491-non-decreasing-subsequences.py
class Solution:
    def findSubsequences(self, nums: List[int]) -> List[List[int]]:
        cur, res = [float("-inf")], set()
        
        def backtrack(idx: int, used: List[bool]) -> None:
            nonlocal res
            if len(cur) >= 3:
                res.add(tuple(cur[1::]))
            for i in range(idx, len(nums)):
                if not used[i] and cur[-1] <= nums[i]:
                    used[i] = True
                    cur.append(nums[i])
                    backtrack(i+1, used)
                    cur.pop()
                    used[i] = False
        
        backtrack(0, [False] * len(nums))
        return res

File: 0491-non-decreasing-subsequences/0491-non-decreasing-subsequences.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var findSubsequences = function(nums) {
    const res = [], map = {}
    backtrack(0, [], Array(nums.length).fill(false))
    return res
    
    function backtrack(idx, list, used) {
        if (list.length >= 2) {
            if (!(list in map)) {
                map[list] = true
                res.push(list.slice())
            }
        }
        for (let i=idx; i < nums.length; i++) {
            if (!used[i] && (!list.length || list.at(-1) <= nums[i])) {
                used[i] = true
                list.push(nums[i])
                backtrack(i+1, list, used)
                list.pop()
                used[i] = false
            }
        }
    }
};